fuel-p2p/src/discovery.rs

```diff
@@ -0,0 +1,507 @@
+                    return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+                        address,
+                        score,
+                    })
+                }
+                NetworkBehaviourAction::CloseConnection {
+                    peer_id,
+                    connection,
+                } => {
+                    return Poll::Ready(NetworkBehaviourAction::CloseConnection {
+                        peer_id,
+                        connection,
+                    })
+                }
+                _ => {}
+            }
+        }
+
+        if let Some(next_event) = self.events.pop_front() {
+            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(next_event));
+        }
+
+        Poll::Pending
+    }
+
+    /// return list of known addresses for a given peer
+    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
+        let mut list = self
+            .bootstrap_nodes
+            .iter()
+            .filter_map(|(current_peer_id, multiaddr)| {
+                if current_peer_id == peer_id {
+                    Some(multiaddr.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        {
+            let mut list_to_filter = Vec::new();
+
+            list_to_filter.extend(self.kademlia.addresses_of_peer(peer_id));
+            list_to_filter.extend(self.mdns.addresses_of_peer(peer_id));
+
+            // filter private addresses
+            // nodes could potentially report addresses in the private network
+            // which are not actually part of the network
+            if !self.allow_private_addresses {
+                list_to_filter.retain(|addr| match addr.iter().next() {
+                    Some(Protocol::Ip4(addr)) if !IpNetwork::from(addr).is_global() => false,
+                    Some(Protocol::Ip6(addr)) if !IpNetwork::from(addr).is_global() => false,
+                    _ => true,
+                });
+            }
+
+            list.extend(list_to_filter);
+        }
+
+        trace!("Addresses of {:?}: {:?}", peer_id, list);
+
+        list
+    }
+
+    fn inject_connection_established(
+        &mut self,
+        peer_id: &PeerId,
+        connection_id: &ConnectionId,
+        endpoint: &libp2p::core::ConnectedPoint,
+        failed_addresses: Option<&Vec<Multiaddr>>,
+    ) {
+        self.connected_peers_count += 1;
+
+        self.kademlia.inject_connection_established(
+            peer_id,
+            connection_id,
+            endpoint,
+            failed_addresses,
+        )
+    }
+
+    fn inject_connected(&mut self, peer_id: &PeerId) {
+        let multiaddr = self.addresses_of_peer(peer_id);
+        self.peer_addresses.insert(*peer_id, multiaddr);
+        self.connected_peers.insert(*peer_id);
+
+        self.events.push_back(DiscoveryEvent::Connected(*peer_id));
+        self.kademlia.inject_connected(peer_id);
```

Comment:
should we include debug logging for these handlers to make troubleshooting easier?

```suggestion
        self.kademlia.inject_connected(peer_id);
        tracing::debug!("Connected to peer {:?}", peer_id)
```

Reply:
yeah we should definitely add more logging